contrib/endpoints/src/api_manager/service_control/aggregated_test.cc

```diff
@@ -195,6 +231,94 @@ TEST_F(AggregatedTestWithRealClient, CheckOKTest) {
   EXPECT_EQ(stat.send_report_operations, 0);
 }
 
+class QuotaAllocationTestWithRealClient : public ::testing::Test {
+ public:
+  void SetUp() {
+    service_.set_name("test_service");
+    service_.mutable_control()->set_environment(
+        "servicecontrol.googleapis.com");
+    env_.reset(new ::testing::NiceMock<MockApiManagerEnvironment>);
+    sc_lib_.reset(Aggregated::Create(service_, nullptr, env_.get(), nullptr));
+    ASSERT_TRUE((bool)(sc_lib_));
+    // This is the call actually creating the client.
+    sc_lib_->Init();
+
+    metric_cost_vector_ = {{"metric_first", 1}, {"metric_second", 2}};
+  }
+
+  void DoRunHTTPRequest(HTTPRequest* request) {
+    std::map<std::string, std::string> headers;
+
+    AllocateQuotaRequest quota_request;
+    AllocateQuotaResponse quota_response;
+
+    ASSERT_TRUE(quota_request.ParseFromString(request->body()));
+    ASSERT_EQ(quota_request.allocate_operation().quota_metrics_size(), 2);
+
+    std::set<std::pair<std::string, int>> expected_costs = {
+        {"metric_first", 1}, {"metric_second", 2}};
+    std::set<std::pair<std::string, int>> actual_costs;
+
+    for (auto rule : quota_request.allocate_operation().quota_metrics()) {
+      actual_costs.insert(std::make_pair(rule.metric_name(),
+                                         rule.metric_values(0).int64_value()));
+    }
+
+    ASSERT_EQ(actual_costs, expected_costs);
+
+    ASSERT_TRUE(::google::protobuf::TextFormat::ParseFromString(
+        kAllocateQuotaResponse, &quota_response));
+
+    std::string body = quota_response.SerializeAsString();
+
+    request->OnComplete(Status::OK, std::move(headers), std::move(body));
+  }
+
+  void DoRunHTTPRequestAllocationFailed(HTTPRequest* request) {
+    std::map<std::string, std::string> headers;
+    AllocateQuotaResponse quota_response;
+
+    ASSERT_TRUE(::google::protobuf::TextFormat::ParseFromString(
+        kAllocateQuotaResponseErrorExhausted, &quota_response));
+
+    std::string body = quota_response.SerializeAsString();
+
```

Comment:
it seems that line 279-284 can be a small function so it can be called by line 269 too.